storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java

```diff
@@ -0,0 +1,204 @@
+        // This can be customized to read remote log segments from followers.
+        readReplicaSelectorClass()
+                .ifPresent(c -> overridingProps.put(KafkaConfig.ReplicaSelectorClassProp(), c.getName()));
+        // The directory of the second-tier storage needs to be constant across all instances of storage managers
+        // in every broker and throughout the test. Indeed, as brokers are restarted during the test.
+        // You can override this property with a fixed path of your choice if you wish to use a non-temporary
+        // directory to access its content after a test terminated.
+        overridingProps.setProperty(storageConfigPrefix(STORAGE_DIR_CONFIG), TestUtils.tempDir().getAbsolutePath());
+        // This configuration will remove all the remote files when close is called in remote storage manager.
+        // Storage manager close is being called while the server is actively processing the socket requests,
+        // so enabling this config can break the existing tests.
+        // NOTE: When using TestUtils#tempDir(), the folder gets deleted when VM terminates.
+        overridingProps.setProperty(storageConfigPrefix(DELETE_ON_CLOSE_CONFIG), "false");
+        return overridingProps;
+    }
+
+    protected Optional<Class<ReplicaSelector>> readReplicaSelectorClass() {
+        return Optional.empty();
+    }
+
+    protected abstract void writeTestSpecifications(TieredStorageTestBuilder builder);
+
+    @BeforeEach
+    @Override
+    public void setUp(TestInfo testInfo) {
+        super.setUp(testInfo);
+        context = new TieredStorageTestContext(brokers(), producerConfig(), consumerConfig(), adminClientConfig());
+    }
+
+    @Disabled("Disabled until the trunk build is stable to test tiered storage")
+    @Test
+    public void executeTieredStorageTest() {
+        TieredStorageTestBuilder builder = new TieredStorageTestBuilder();
+        writeTestSpecifications(builder);
+        try {
+            for (TieredStorageTestAction action : builder.complete()) {
+                action.execute(context);
+            }
+        } catch (Exception ex) {
+            throw new AssertionError("Could not build test specifications. No test was executed.", ex);
+        }
+    }
+
+    @AfterEach
+    @Override
+    public void tearDown() {
+        try {
+            context.close();
+            super.tearDown();
+            context.printReport(System.out);
+        } catch (Exception ex) {
+            throw new AssertionError("Failed to close the tear down the test harness.", ex);
+        }
+    }
+
+    private String storageConfigPrefix(String key) {
+        return LocalTieredStorage.STORAGE_CONFIG_PREFIX + key;
+    }
+
+    private String metadataConfigPrefix(String key) {
+        return "rlmm.config." + key;
+    }
+
+    @SuppressWarnings("deprecation")
+    public static List<LocalTieredStorage> getTieredStorages(Seq<KafkaBroker> brokers) {
```

Comment:
getTieredStorages is confusing here since this method gives an instance of RSM plugins for each broker. Could we rename this to `remoteStorageManagers()`?